README.md
[![Python Versions](https://img.shields.io/pypi/pyversions/digitalearth.png)](https://img.shields.io/pypi/pyversions/digitalearth)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/MAfarrag/digitalearth.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/MAfarrag/digitalearth/context:python)
[![Documentation Status](https://readthedocs.org/projects/digitalearth/badge/?version=latest)](https://cleopatra.readthedocs.io/en/latest/?badge=latest)


[![codecov](https://codecov.io/gh/Serapieum-of-alex/Digital-Earth/branch/main/graph/badge.svg?token=nDBDBjsyij)](https://codecov.io/gh/Serapieum-of-alex/Digital-Earth)
![GitHub last commit](https://img.shields.io/github/last-commit/MAfarrag/digitalearth)
![GitHub forks](https://img.shields.io/github/forks/MAfarrag/digitalearth?style=social)
![GitHub Repo stars](https://img.shields.io/github/stars/MAfarrag/digitalearth?style=social)


Current release info
====================

| Name | Downloads                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | Version | Platforms |
| --- |------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| --- | --- |
| [![Conda Recipe](https://img.shields.io/badge/recipe-digitalearth-green.svg)](https://anaconda.org/conda-forge/digitalearth) | [![Conda Downloads](https://img.shields.io/conda/dn/conda-forge/digitalearth.svg)](https://anaconda.org/conda-forge/digitalearth) [![Downloads](https://pepy.tech/badge/digitalearth)](https://pepy.tech/project/digitalearth) [![Downloads](https://pepy.tech/badge/digitalearth/month)](https://pepy.tech/project/digitalearth)  [![Downloads](https://pepy.tech/badge/digitalearth/week)](https://pepy.tech/project/digitalearth)  ![PyPI - Downloads](https://img.shields.io/pypi/dd/digitalearth?color=blue&style=flat-square) ![GitHub all releases](https://img.shields.io/github/downloads/MAfarrag/digitalearth/total) ![GitHub release (latest by date)](https://img.shields.io/github/downloads/MAfarrag/digitalearth/0.1.0/total) | [![Conda Version](https://img.shields.io/conda/vn/conda-forge/digitalearth.svg)](https://anaconda.org/conda-forge/digitalearth) [![PyPI version](https://badge.fury.io/py/digitalearth.svg)](https://badge.fury.io/py/digitalearth) [![Anaconda-Server Badge](https://anaconda.org/conda-forge/digitalearth/badges/version.svg)](https://anaconda.org/conda-forge/digitalearth) | [![Conda Platforms](https://img.shields.io/conda/pn/conda-forge/digitalearth.svg)](https://anaconda.org/conda-forge/digitalearth) [![Join the chat at https://gitter.im/Hapi-Nile/Hapi](https://badges.gitter.im/Hapi-Nile/Hapi.svg)](https://gitter.im/Hapi-Nile/Hapi?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) |

digitalearth - Remote Sensing package
=====================================================================
**digitalearth** is a Remote Sensing package

digitalearth

Main Features
-------------
  - plot static maps


Future work
-------------
  - dynamic/interactive maps



Installing digitalearth
===============

Installing `digitalearth` from the `conda-forge` channel can be achieved by:

```
conda install -c conda-forge digitalearth
```

It is possible to list all of the versions of `digitalearth` available on your platform with:

```
conda search digitalearth --channel conda-forge
```

## Install from Github
to install the last development to time you can install the library from github
```
pip install git+https://github.com/MAfarrag/digitalearth
```

## pip
to install the last release you can easly use pip
```
pip install digitalearth==0.1.11
```

Quick start
===========

```
from digitalearth.map import Map
src = gdal.Open("examples/data/acc4000.tif")
fig, ax = Map.plot(src, title="Flow Accumulation", cbar_label="Flow Accumulation")  
```
![Flowaccumulation](examples/images/flow_accumulation.png)
```
points = pd.read_csv("examples/data/points.csv")
point_color = "blue"
point_size = 100
id_color = "yellow"
id_size = 20

display_cellvalue = True
num_size = 8
background_color_threshold = None
ticks_spacing = 500

fig, ax = Map.plot(
            src,
            point_color=point_color,
            point_size=point_size,
            pid_color=id_color,
            pid_size=id_size,
            points=points,
            display_cellvalue=display_cellvalue,
            num_size=num_size,
            background_color_threshold=background_color_threshold,
            ticks_spacing=ticks_spacing,
            title="Flow Accumulation",
            cbar_label="Flow Accumulation"
        )
```
![Flowaccumulation](examples/images/flow_accumulation_with_labels.png)
[other code samples](https://digitalearth.readthedocs.io/en/latest/?badge=latest)


HISTORY.rst
=======
History
=======

0.1.0 (2022-05-24)
------------------

* First release on PyPI.

0.1.8 (2022-12-19)
------------------

* Use environment.yaml and requirements.txt instead of pyproject.toml and replace poetry env byconda env

0.1.9 (2022-12-19)
------------------

* lock numpy to version 1.23.5

0.1.10 (2022-12-27)
------------------

* fix pypi package names in the requirements.txt
* fix python version number

0.1.11 (2023-01-31)
------------------

* bump up cleopatra version
* bump up pyramids version
* add github action for creating a release from tags


tests/mo_plot.py
import matplotlib
matplotlib.use("TkAgg")
# import matplotlib.pyplot as plt
from digitalearth.map import Map
from osgeo import gdal

#%%
src = gdal.Open("examples/data/acc4000.tif")
src_arr = src.ReadAsArray()
src_no_data_value = src.GetRasterBand(1).GetNoDataValue()
cmap = "terrain"
#%%
fig, ax = Map.plotGeo(src, Title="Flow Accumulation")
#%%
fig, ax = Map.plotArray(src_arr, nodataval=src_no_data_value, Title="Flow Accumulation")

#%%



tests/__init__.py
def raster():
    return None


tests/conftest.py
from typing import List

import numpy as np
import pandas as pd
import pytest
from osgeo import gdal
from osgeo.gdal import Dataset


@pytest.fixture(scope="module")
def src() -> Dataset:
    return gdal.Open("examples/data/acc4000.tif")


@pytest.fixture(scope="module")
def display_cellvalue() -> bool:
    return True


@pytest.fixture(scope="module")
def background_color_threshold():
    return None


@pytest.fixture(scope="module")
def num_size() -> int:
    return 8


@pytest.fixture(scope="module")
def ticks_spacing() -> int:
    return 500


@pytest.fixture(scope="module")
def points() -> pd.DataFrame:
    return pd.read_csv("examples/data/points.csv")


@pytest.fixture(scope="module")
def pid_size() -> int:
    return 20


@pytest.fixture(scope="module")
def pid_color() -> str:
    return "green"


@pytest.fixture(scope="module")
def point_size() -> int:
    return 100


@pytest.fixture(scope="module")
def point_color() -> str:
    return "blue"


digitalearth/map.py
from typing import Any, List, Tuple, Union

# import geoplot as gplt
# import geoplot.crs as gcrs
# import matplotlib.pyplot as plt
import numpy as np
from cleopatra.array import Array
# from geopandas import GeoDataFrame
# from loguru import logger
from osgeo.gdal import Dataset
from pyramids.catchment import Catchment as GC
from pyramids.raster import Raster

# from Hapi.plot.visualizer import MidpointNormalize, Map



class Map:
    """
    Map
    """
    figure_default_options = dict(
        ylabel="",
        xlabel="",
        legend="",
        legend_size=10,
        figsize=(10, 8),
        labelsize=10,
        fontsize=10,
        name="hist.tif",
        color1="#3D59AB",
        color2="#DC143C",
        linewidth=3,
        Axisfontsize=15,
    )


    def __init__(self):
        pass


    @staticmethod
    def plot(
            src: Union[Dataset, np.ndarray],
            point_color: str = "red",
            point_size: Union[int, float] = 100,
            pid_color="blue",
            pid_size: Union[int, float] = 10,
            **kwargs
    ):
        """plot.

            plot an array/ gdal dataset

        Parameters
        ----------
        src : [array/gdal.Dataset]
            the array/gdal raster you want to plot.
        point_color : [str], optional
            color of the points. The default is 'red'.
        point_size : [integer], optional
            size of the points. The default is 100.
        pid_color : [str]
            the ID of the Point.The default is "blue".
        pid_size : [integer]
            size of the ID text. The default is 10.
        pid_color : []


        **kwargs : [dict]
            keys:
                nodataval: Union[int, float] = np.nan,
                figsize: Tuple[int, int] = (8, 8),
                title: Any = "Total Discharge",
                title_size: Union[int, float] = 15,
                cbar_length: Union[int, float] = 0.75,
                orientation: str = "vertical",
                cbar_label_size: Union[int, float] = 12,
                cbar_label: str = "Color bar label",
                rotation: Union[int, float] = -90,
                ticks_spacing: Union[int, float] = 5,
                num_size: Union[int, float] = 8,
                color_scale: int = 1,
                cmap: str = "coolwarm_r",
                gamma: Union[int, float] = 0.5,
                linscale: Union[int, float] = 0.001,
                linthresh: Union[int, float] = 0.0001,
                midpoint: int = 0,
                display_cellvalue: bool = False,
                background_color_threshold=None,

        Returns
        -------
        axes: [figure axes].
            the axes of the matplotlib figure
        fig: [matplotlib figure object]
            the figure object
        """
        if isinstance(src, Dataset):
            arr, nodataval = Raster.getRasterData(src)
        else:
            arr = src
            if "nodataval" not in kwargs.keys():
                raise ValueError("If the first parameter is a numpy.ndarray object you have to enter a kwargs 'nodataval'"
                                 "value")
            else:
                nodataval = kwargs["nodataval"]
        # convert the array to float as integer array gives error when compared to float
        arr = arr.astype(np.float32)

        if nodataval is not None:
            arr[np.isclose(arr, nodataval, rtol=0.001)] = np.nan

        if "points" in kwargs.keys():
            points = kwargs["points"]
            points["row"] = np.nan
            points["col"] = np.nan
            # to locte the points in the array
            points.loc[:, ["row", "col"]] = GC.nearestCell(
                src, points[["x", "y"]][:]
            ).values

        fig, ax = Array.plot(arr, nodataval, **kwargs)

        PoitsID = list()
        if "points" in kwargs.keys():
            row = points.loc[:, "row"].tolist()
            col = points.loc[:, "col"].tolist()
            IDs = points.loc[:, "id"].tolist()
            Points = ax.scatter(col, row, color=point_color, s=point_size)

            for i in range(len(row)):
                PoitsID.append(
                    ax.text(
                        col[i],
                        row[i],
                        IDs[i],
                        ha="center",
                        va="center",
                        color=pid_color,
                        fontsize=pid_size,
                    )
                )

        return fig, ax


    # @staticmethod
    # def plotCatchment(
    #         points: GeoDataFrame,
    #         column_name: Any,
    #         poly: GeoDataFrame,
    #         line: GeoDataFrame,
    #         scheme: Any = None,
    #         scale_func: Any = '',
    #         cmap: str = "viridis",
    #         legend_values: List = [],
    #         legend_labels: List = [],
    #         figsize: Tuple = (8, 8),
    #         title: Any = 'title',
    #         title_size: int = 500,
    #         save: Union[bool, str] = False,
    # ):
    #     """PlotCatchment.
    #
    #     Inputs:
    #     ------
    #     points:[GeoDataFrame]
    #         geodataframe contains values to plot in one of its columns.
    #     column_name: [str]
    #         name of the column you want to plot its values.
    #     poly: [GeoDataFrame]
    #         geodataframe contains polygon geometries.
    #     line: [GeoDataFrame]
    #         geodataframe contains linestring geometries.
    #     figsize: [Tuple]
    #         fize oif the figure.
    #     title:[str]
    #         title of the figure.
    #     save: [bool/str]
    #         if you want to save the plot provide the path with the extention,
    #         Default is False.
    #     """
    #     # unify the projection
    #     if not poly.crs.is_geographic:
    #         logger.debug("The coordinate system of the poly geodataframe is not geographic"
    #                      "SO, it will be reprojected to WGS-84")
    #         poly.to_crs(4326, inplace=True)
    #
    #     epsg = poly.crs.to_json()
    #     line.to_crs(epsg, inplace=True)
    #     points.to_crs(epsg, inplace=True)
    #
    #     pointplot_kwargs = {'edgecolor': 'white', 'linewidth': 0.9}  # 'color': "crimson"
    #
    #     # make sure that the plotted column is numeric
    #     points[column_name] = points[column_name].map(float)
    #
    #     fig, ax = plt.subplots(1, 1, figsize=figsize, subplot_kw={'projection': gcrs.AlbersEqualArea()})
    #     if scheme:
    #
    #         gplt.pointplot(points, projection=gcrs.AlbersEqualArea(),
    #                        hue=column_name, cmap=cmap,
    #                        scale=column_name, limits=(4, 20),
    #                        scheme=scheme,
    #                        # scale_func = scale_func,
    #                        legend=True,
    #                        legend_var='scale',
    #                        legend_kwargs={  # 'loc': 'upper right',
    #                            'bbox_to_anchor': (1, 0.35)},
    #                        ax=ax, **pointplot_kwargs  # ,
    #                        )
    #     else:
    #         gplt.pointplot(points, projection=gcrs.AlbersEqualArea(),
    #                        hue=column_name, cmap=cmap,
    #                        scale=column_name, limits=(4, 20),
    #                        scale_func=scale_func,
    #                        legend=True,
    #                        legend_var='scale',
    #                        legend_values=legend_values,
    #                        legend_labels=legend_labels,
    #                        legend_kwargs={  # 'loc': 'upper right',
    #                            'bbox_to_anchor': (1, 0.35)},
    #                        ax=ax, **pointplot_kwargs  # ,
    #                        )
    #
    #     gplt.polyplot(poly, ax=ax, edgecolor='grey', facecolor='grey',  # 'lightgray',
    #                   linewidth=0.5, extent=poly.total_bounds)  # # , zorder=0
    #
    #     gplt.polyplot(line, ax=ax, linewidth=10)
    #
    #     plt.title(title, fontsize=title_size)
    #     # plt.subplots_adjust(top=0.99999, right=0.9999, left=0.000005, bottom=0.000005)
    #     if save:
    #         plt.savefig(save, bbox_inches='tight', transparent=True)
    #
    #     return fig, ax


digitalearth/__init__.py
try:
    from importlib.metadata import PackageNotFoundError  # type: ignore
    from importlib.metadata import version
except ImportError:  # pragma: no cover
    from importlib_metadata import PackageNotFoundError  # type: ignore
    from importlib_metadata import version


try:
    __version__ = version(__name__)
except PackageNotFoundError:  # pragma: no cover
    __version__ = "unknown"

# documentation format
__author__ = "Mostafa Farrag"
__email__ = 'moah.farag@gmail.com'
__docformat__ = "restructuredtext"

# Let users know if they're missing any of our hard dependencies
hard_dependencies = ()  # ("numpy", "pandas", "gdal")
missing_dependencies = []

for dependency in hard_dependencies:
    try:
        __import__(dependency)
    except ImportError as e:
        missing_dependencies.append(dependency)

if missing_dependencies:
    raise ImportError("Missing required dependencies {0}".format(missing_dependencies))


import digitalearth.static

__doc__ = """
digitalearth - visualization package
"""


examples/plot_examples.py
import matplotlib

matplotlib.use("TkAgg")
import geopandas as gpd
import mapclassify as mc

# import matplotlib.pyplot as plt
import pandas as pd
from osgeo import gdal

from digitalearth.static import Map

saveto = r"examples\data\results"
#%%
src = gdal.Open("examples/data/acc4000.tif")
src_no_data_value = src.GetRasterBand(1).GetNoDataValue()
cmap = "terrain"
#%%
fig, ax = Map.plot(src, title="Flow Accumulation", cbar_label="Flow Accumulation")
#%%
points = pd.read_csv("examples/data/points.csv")
# fig, ax = Map.plot(src, title="Flow Accumulation", points=points)

point_color = "blue"
point_size = 100
id_color = "yellow"
id_size = 20

display_cellvalue = True
num_size = 8
background_color_threshold = None
ticks_spacing = 500

fig, ax = Map.plot(
    src,
    point_color=point_color,
    point_size=point_size,
    pid_color=id_color,
    pid_size=id_size,
    points=points,
    display_cellvalue=display_cellvalue,
    num_size=num_size,
    background_color_threshold=background_color_threshold,
    ticks_spacing=ticks_spacing,
    title="Flow Accumulation",
    cbar_label="Flow Accumulation",
)
#%%
# read the polygon , line, and point data
rhine_basin = gpd.read_file("examples/data/rhine_basin.geojson")
print(rhine_basin.loc[0, "geometry"].geom_type)
rhine_basin.plot()
rhine_river = gpd.read_file("examples/data/rhine_river_centerline.geojson")
print(rhine_river.loc[0, "geometry"].geom_type)
rhine_river.plot()
metrix = gpd.read_file("examples/data/MetricsHM_Q_Obs.geojson")
print(metrix.loc[0, "geometry"].geom_type)
metrix.plot()

"The columns that we can plot its values in the dataframe"
print(metrix.columns)
"Check the projection"
print(rhine_river.crs)
print(rhine_basin.crs)
print(metrix.crs)


plot_column = "NSE"

metrix[plot_column] = metrix[plot_column].map(float)
"check the histogram for the values that you are going to plot using the bubble plot"
metrix.loc[:, plot_column].plot.hist(bins=7)
"plot "
ax = metrix.plot(column=plot_column)
"plot the values against the index"
ax = metrix.loc[:, plot_column].plot()
ax.set_xlabel("")
ax.set_ylabel("NSE")


filterMin = None
filterMax = None

# filter some values out
if filterMin:
    metrix.loc[metrix[plot_column] < filterMin, plot_column] = filterMin
if filterMax:
    metrix.loc[metrix[plot_column] > filterMax, plot_column] = filterMax
#%% apply a size function your self
"apply the function that calculates a size based on the values in the column"


def scale_func(x):
    if x >= 1:
        return 20
    elif 0.6 <= x < 0.8:
        return 14
    elif 0.4 <= x < 0.6:
        return 10
    elif 0.2 <= x < 0.4:
        return 6
    elif 0.1 <= x < 0.2:
        return 4
    else:
        return 1


title = f"Apply function-{plot_column}"
new_col = f"{plot_column}-size"
metrix[new_col] = metrix[plot_column].apply(scale_func)

legend_values = [20, 14, 10, 6, 4, 1]
legend_labels = [1, 0.8, 0.6, 0.4, 0.2, 0.1]

cmap = "Blues"
linewidth = 0.2

fig, ax = Map.plotCatchment(
    metrix,
    new_col,
    rhine_basin,
    rhine_river,
    title=title,
    title_size=15,
    legend_values=legend_values,
    legend_labels=legend_labels,
    cmap=cmap,
    linewidth=linewidth,
    save=rf"{saveto}\{title}.tif",
)

#%% create a color classifier using map classify
scheme = mc.Quantiles(metrix[plot_column].map(float), k=3)
title = f"Quantiles-{plot_column}"
cmap = "Blues"
linewidth = 0.2

fig, ax = Map.plotCatchment(
    metrix,
    plot_column,
    rhine_basin,
    rhine_river,
    title=title,
    title_size=15,
    scheme=scheme,
    cmap=cmap,
    linewidth=linewidth,
    save=rf"{saveto}\{title}.tif",
)
#%%
title = f"Enter scale function-{plot_column}"
scale_func = scale_func
# legend_values = [1, 0.8, 0.6, 0.4, 0.2, 0.1, 0]

cmap = "Blues"
linewidth = 0.2

fig, ax = Map.plotCatchment(
    metrix,
    plot_column,
    rhine_basin,
    rhine_river,
    title=title,
    title_size=15,
    scheme=scheme,
    scale_func=scale_func,
    # legend_values = legend_values,
    # legend_labels = legend_values,
    cmap=cmap,
    linewidth=linewidth,
    save=rf"{saveto}\{title}.tif",
)


digitalearth/static.py
from typing import Any, List, Tuple, Union

import matplotlib.pyplot as plt
import numpy as np
from cleopatra.array import Array
from geopandas import GeoDataFrame
from loguru import logger
from osgeo.gdal import Dataset
from pyramids.catchment import Catchment as GC
from pyramids.raster import Raster

# from Hapi.plot.visualizer import MidpointNormalize, Map


class Map:
    """Map."""

    figure_default_options = dict(
        ylabel="",
        xlabel="",
        legend="",
        legend_size=10,
        figsize=(10, 8),
        labelsize=10,
        fontsize=10,
        name="hist.tif",
        color1="#3D59AB",
        color2="#DC143C",
        linewidth=3,
        Axisfontsize=15,
    )

    def __init__(self):
        pass

    @staticmethod
    def plot(
        src: Union[Dataset, np.ndarray],
        point_color: str = "red",
        point_size: Union[int, float] = 100,
        pid_color="blue",
        pid_size: Union[int, float] = 10,
        **kwargs
    ):
        """plot.

            plot an array/ gdal dataset

        Parameters
        ----------
        src : [array/gdal.Dataset]
            the array/gdal raster you want to plot.
        point_color : [str], optional
            color of the points. The default is 'red'.
        point_size : [integer], optional
            size of the points. The default is 100.
        pid_color : [str]
            the ID of the Point.The default is "blue".
        pid_size : [integer]
            size of the ID text. The default is 10.
        pid_color : []


        **kwargs : [dict]
            keys:
                nodataval: Union[int, float] = np.nan,
                figsize: Tuple[int, int] = (8, 8),
                title: Any = "Total Discharge",
                title_size: Union[int, float] = 15,
                cbar_length: Union[int, float] = 0.75,
                orientation: str = "vertical",
                cbar_label_size: Union[int, float] = 12,
                cbar_label: str = "Color bar label",
                rotation: Union[int, float] = -90,
                ticks_spacing: Union[int, float] = 5,
                num_size: Union[int, float] = 8,
                color_scale: int = 1,
                cmap: str = "coolwarm_r",
                gamma: Union[int, float] = 0.5,
                linscale: Union[int, float] = 0.001,
                linthresh: Union[int, float] = 0.0001,
                midpoint: int = 0,
                display_cellvalue: bool = False,
                background_color_threshold=None,

        Returns
        -------
        axes: [figure axes].
            the axes of the matplotlib figure
        fig: [matplotlib figure object]
            the figure object
        """
        if isinstance(src, Dataset):
            arr, nodataval = Raster.getRasterData(src)
        else:
            arr = src
            if "nodataval" not in kwargs.keys():
                raise ValueError(
                    "If the first parameter is a numpy.ndarray object you have to enter a kwargs 'nodataval'"
                    "value"
                )
            else:
                nodataval = kwargs["nodataval"]
        # convert the array to float as integer array gives error when compared to float
        arr = arr.astype(np.float32)

        if nodataval is not None:
            arr[np.isclose(arr, nodataval, rtol=0.001)] = np.nan

        if "points" in kwargs.keys():
            points = kwargs["points"]
            points["row"] = np.nan
            points["col"] = np.nan
            # to locte the points in the array
            points.loc[:, ["row", "col"]] = GC.nearestCell(
                src, points[["x", "y"]][:]
            ).values

        fig, ax = Array.plot(arr, nodataval, **kwargs)

        points_ids = list()
        if "points" in kwargs.keys():
            row = points.loc[:, "row"].tolist()
            col = points.loc[:, "col"].tolist()
            IDs = points.loc[:, "id"].tolist()
            ax.scatter(col, row, color=point_color, s=point_size)
            # TODO: Points = ax.scatter(col, row, color=point_color, s=point_size)
            #  return the scatter plot object (Points)

            for i in range(len(row)):
                points_ids.append(
                    ax.text(
                        col[i],
                        row[i],
                        IDs[i],
                        ha="center",
                        va="center",
                        color=pid_color,
                        fontsize=pid_size,
                    )
                )

        return fig, ax

    @staticmethod
    def plotCatchment(
        points: GeoDataFrame,
        column_name: Any,
        poly: GeoDataFrame,
        line: GeoDataFrame,
        scheme: Any = None,
        scale_func: Any = None,
        cmap: str = "viridis",
        legend_values: List = [],
        legend_labels: List = [],
        figsize: Tuple = (8, 8),
        title: Any = "title",
        title_size: int = 500,
        linewidth: float = 0.5,
        save: Union[bool, str] = False,
    ):
        """PlotCatchment.

        Parameters
        ----------
        points:[GeoDataFrame]
            geodataframe contains values to plot in one of its columns.
        column_name: [str]
            name of the column you want to plot its values.
        poly: [GeoDataFrame]
            geodataframe contains polygon geometries.
        line: [GeoDataFrame]
            geodataframe contains linestring geometries.
        linewidth:
        title_size:
        legend_labels:
        legend_values:
        cmap:
        scale_func:
        scheme:
        figsize: [Tuple]
            fize oif the figure.
        title:[str]
            title of the figure.
        save: [bool/str]
            if you want to save the plot provide the path with the extention,
            Default is False.
        """
        import geoplot as gplt
        import geoplot.crs as gcrs

        # unify the projection
        if not poly.crs.is_geographic:
            logger.debug(
                "The coordinate system of the poly geodataframe is not geographic "
                "SO, it will be reprojected to WGS-84"
            )
            poly.to_crs(4326, inplace=True)

        epsg = poly.crs.to_json()
        line.to_crs(epsg, inplace=True)
        points.to_crs(epsg, inplace=True)

        pointplot_kwargs = {
            "edgecolor": "white",
            "linewidth": 0.9,
        }  # 'color': "crimson"

        # make sure that the plotted column is numeric
        points[column_name] = points[column_name].map(float)

        fig, ax = plt.subplots(
            1, 1, figsize=figsize, subplot_kw={"projection": gcrs.AlbersEqualArea()}
        )

        if scheme:
            gplt.pointplot(
                points,
                projection=gcrs.AlbersEqualArea(),
                hue=column_name,
                cmap=cmap,
                scale=column_name,
                limits=(4, 20),
                scheme=scheme,
                legend=True,
                legend_var="scale",
                legend_kwargs={"bbox_to_anchor": (1, 0.35)},  # 'loc': 'upper right',
                ax=ax,
                **pointplot_kwargs  # ,
            )
        else:
            if scale_func:
                gplt.pointplot(
                    points,
                    projection=gcrs.AlbersEqualArea(),
                    hue=column_name,
                    cmap=cmap,
                    scale=column_name,
                    limits=(4, 20),
                    scale_func=scale_func,
                    legend=True,
                    legend_var="scale",
                    legend_values=legend_values,
                    legend_labels=legend_labels,
                    legend_kwargs={  # 'loc': 'upper right',
                        "bbox_to_anchor": (1, 0.35)
                    },
                    ax=ax,
                    **pointplot_kwargs  # ,
                )
            else:
                gplt.pointplot(
                    points,
                    projection=gcrs.AlbersEqualArea(),
                    hue=column_name,
                    cmap=cmap,
                    scale=column_name,
                    limits=(4, 20),
                    # scale_func=scale_func,
                    legend=True,
                    legend_var="scale",
                    legend_values=legend_values,
                    legend_labels=legend_labels,
                    legend_kwargs={  # 'loc': 'upper right',
                        "bbox_to_anchor": (1, 0.35)
                    },
                    ax=ax,
                    **pointplot_kwargs  # ,
                )

        gplt.polyplot(
            poly,
            ax=ax,
            edgecolor="grey",
            facecolor="grey",  # 'lightgray',
            linewidth=0.5,
            extent=poly.total_bounds,
        )  # # , zorder=0

        gplt.polyplot(line, ax=ax, linewidth=10)

        plt.title(title, fontsize=title_size)
        # plt.subplots_adjust(top=0.99999, right=0.9999, left=0.000005, bottom=0.000005)
        if save:
            plt.savefig(save, bbox_inches="tight", transparent=True)

        return fig, ax
